src/requirementslib/models/utils.py

```diff
@@ -149,3 +156,326 @@ def validate_specifiers(instance, attr_, value):
+
+    :param name: The requirement's name.
+    :type name: str
+    :param version: The requirement version (must be pinned).
+    :type version: str.
+    :param extras: The desired extras.
+    :type extras: list[str]
+    :param markers: The desired markers, without a preceding semicolon.
+    :type markers: str
+    :param constraint: Whether to flag the requirement as a constraint, defaults to False.
+    :param constraint: bool, optional
+    :return: A generated InstallRequirement
+    :rtype: :class:`~pip._internal.req.req_install.InstallRequirement`
+    """
+
+    # If no extras are specified, the extras string is blank
+    extras_string = ""
+    if extras:
+        # Sort extras for stability
+        extras_string = "[{}]".format(",".join(sorted(extras)))
+
+    if not markers:
+        return InstallRequirement.from_line(
+            str('{}{}=={}'.format(name, extras_string, version)),
+            constraint=constraint)
+    else:
+        return InstallRequirement.from_line(
+            str('{}{}=={}; {}'.format(name, extras_string, version, str(markers))),
+            constraint=constraint)
+
+
+def version_from_ireq(ireq):
+    """version_from_ireq Extract the version from a supplied :class:`~pip._internal.req.req_install.InstallRequirement`
+
+    :param ireq: An InstallRequirement
+    :type ireq: :class:`~pip._internal.req.req_install.InstallRequirement`
+    :return: The version of the InstallRequirement.
+    :rtype: str
+    """
+
+    return first(ireq.specifier._specs).version
+
+
+def partialclass(cls, *args, **kwds):
+    """Returns a partially instantiated class
+
+    :return: A partial class instance
+    :rtype:
+    """
+
+    class NewCls(cls):
+        __init__ = partialmethod(cls.__init__, *args, **kwds)
+
+    return NewCls
+
+
+def clean_requires_python(candidates):
+    """Get a cleaned list of all the candidates with valid specifiers in the `requires_python` attributes."""
+    all_candidates = []
+    sys_version = '.'.join(map(str, sys.version_info[:3]))
+    py_version = parse_version(os.environ.get('PIP_PYTHON_VERSION', sys_version))
+    for c in candidates:
+        from_location = attrgetter("location.requires_python")
+        requires_python = getattr(c, "requires_python", from_location(c))
+        if requires_python:
+            # Old specifications had people setting this to single digits
+            # which is effectively the same as '>=digit,<digit+1'
+            if requires_python.isdigit():
+                requires_python = '>={0},<{1}'.format(requires_python, int(requires_python) + 1)
+            try:
+                specifierset = SpecifierSet(requires_python)
+            except InvalidSpecifier:
+                continue
+            else:
+                if not specifierset.contains(py_version):
+                    continue
+        all_candidates.append(c)
+    return all_candidates
+
+
+def fix_requires_python_marker(requires_python):
```

Comment:
the thing to note about this is that technically, in any case where you would "and" two markers which have the same variable and which are numeric (i.e. versions), you can also represent them in a comma-separated list without the and